File: arrayChallenges/moveZeroes.c
#include <stdio.h>

int optimizeLeadingZeroes(int* nums,int numsSize){
		for(int i=numsSize-1;i>0;i--){
				if (*(nums+i) != 0) {
						return i+1;
				}
		}
		return 0;
}

void moveZeroes(int* nums,int numsSize){
		int end,shift = 0;
		end = optimizeLeadingZeroes(nums,numsSize);
		printf("%d\n",end);
		for(int i=0;i<end;i++){
				if (*(nums + i) == 0){
						shift++;
				}
				else if (shift > 0){
						*(nums + i - shift) = *(nums + i);
						*(nums + i) = 0;
				}
				printf("\n%d\t%d\n",i,shift);
				for (int i=0;i<9;i++) {
						printf("%d\n",nums[i]);
				}
		}
}






int main() {
		int inums[9] = {1,2,0,4,5,6,7,8,0};
		moveZeroes(inums,9);
		return 0;
}

File: arrayChallenges/moveZeroes.py
"""This solution contains an optimization regarding leading zeroes, so it will only loop through numbers until the leading zeroes"""


def moveZeroes(nums) -> None:
    iterator = 0
    shift = 0
    end = 0
    for i in range(len(nums)-1,0,-1):
        if nums[i] != 0:
            end = i
            break

    while iterator <= end:
        if nums[iterator] == 0:
            shift += 1 
        elif shift > 0:
            nums[iterator-shift] = nums[iterator]
            nums[iterator] = 0
        iterator += 1






def main():
    tests = {1: [1,2,3,0,5,0,7,0,0,2,3,0],
             2:[1,2,0,0,0],
             3:[0,0,1,2,3,0,4],
             4:[0,0,0,0],
             5:[0,1,0]}

    for key in tests.keys():
        moveZeroes(tests[key])
        print(tests[key])

if __name__ == "__main__":
    main()

